Auth Failures are not published to the analytics portal.

### Description:
If the authentication failure happens when an exception is thrown (Unclassified Authentication Failure), analytics event is not published.

Here is a sample stack trace of such auth failure.

```
SEVERE: Exception while executing runnable io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed@259e9e33
enforcer_1  | java.lang.SecurityException: Not a JWT token. Failed to decode the token header.
enforcer_1  | 	at org.wso2.choreo.connect.enforcer.security.jwt.JWTAuthenticator.authenticate(JWTAuthenticator.java:108)
enforcer_1  | 	at org.wso2.choreo.connect.enforcer.security.AuthFilter.authenticate(AuthFilter.java:123)
enforcer_1  | 	at org.wso2.choreo.connect.enforcer.security.AuthFilter.handleRequest(AuthFilter.java:105)
enforcer_1  | 	at org.wso2.choreo.connect.enforcer.api.API.executeFilterChain(API.java:41)
enforcer_1  | 	at org.wso2.choreo.connect.enforcer.api.RestAPI.process(RestAPI.java:99)
enforcer_1  | 	at org.wso2.choreo.connect.enforcer.server.RequestHandler.process(RequestHandler.java:60)
enforcer_1  | 	at org.wso2.choreo.connect.enforcer.grpc.ExtAuthService.check(ExtAuthService.java:50)
enforcer_1  | 	at io.envoyproxy.envoy.service.auth.v3.AuthorizationGrpc$MethodHandlers.invoke(AuthorizationGrpc.java:245)
enforcer_1  | 	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:180)
enforcer_1  | 	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:331)
enforcer_1  | 	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:814)
enforcer_1  | 	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
enforcer_1  | 	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
enforcer_1  | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
enforcer_1  | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
enforcer_1  | 	at java.base/java.lang.Thread.run(Unknown Source)
enforcer_1  | Caused by: java.text.ParseException: Invalid JWS header: Invalid JSON: Unexpected token 
enforcer_1  | 	at com.nimbusds.jose.JWSObject.<init>(JWSObject.java:155)
enforcer_1  | 	at com.nimbusds.jwt.SignedJWT.<init>(SignedJWT.java:78)
enforcer_1  | 	at com.nimbusds.jwt.SignedJWT.parse(SignedJWT.java:116)
enforcer_1  | 	at org.wso2.choreo.connect.enforcer.security.jwt.JWTAuthenticator.getSignedJwt(JWTAuthenticator.java:585)
enforcer_1  | 	at org.wso2.choreo.connect.enforcer.security.jwt.JWTAuthenticator.authenticate(JWTAuthenticator.java:106)
enforcer_1  | 	... 15 more
enforcer_1  | 

```

### Steps to reproduce:
Provide an invalid token such that it is not a JWT token format.

### Affected Product Version:
choreo-connect-0.9.0 alpha

### Environment details (with versions):
- OS:
- Client:
- Env (Docker/K8s):

---
### Optional Fields
#### Related Issues:
<!-- Any related issues from this/other repositories-->

#### Suggested Labels:
<!--Only to be used by non-members-->

#### Suggested Assignees:
<!--Only to be used by non-members-->

--- a/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/exception/APISecurityException.java
+++ b/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/exception/APISecurityException.java
@@ -33,8 +33,9 @@ public class APISecurityException extends Exception {
         this.errorCode = errorCode;
     }
 
-    public APISecurityException(int errorCode, String message, Throwable cause) {
+    public APISecurityException(int statusCode, int errorCode, String message, Throwable cause) {
         super(message, cause);
+        this.statusCode = statusCode;
         this.errorCode = errorCode;
     }
 

--- a/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java
+++ b/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java
@@ -108,7 +108,7 @@ public class JWTAuthenticator implements Authenticator {
             log.error("Failed to decode the token header", e);
             throw new APISecurityException(APIConstants.StatusCodes.UNAUTHENTICATED.getCode(),
                     APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
-                    "Not a JWT token. Failed to decode the token header.", e);
+                    "Not a JWT token. Failed to decode the token header", e);
         }
         JWTClaimsSet claims = signedJWTInfo.getJwtClaimsSet();
         String jwtTokenIdentifier = getJWTTokenIdentifier(signedJWTInfo);

--- a/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java
+++ b/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java
@@ -107,7 +107,7 @@ public class JWTAuthenticator implements Authenticator {
         } catch (ParseException | IllegalArgumentException e) {
             log.error("Failed to decode the token header", e);
             throw new APISecurityException(APIConstants.StatusCodes.UNAUTHENTICATED.getCode(),
-                    APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                    APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
                     "Not a JWT token. Failed to decode the token header.", e);
         }
         JWTClaimsSet claims = signedJWTInfo.getJwtClaimsSet();

--- a/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java
+++ b/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java
@@ -101,12 +101,14 @@ public class JWTAuthenticator implements Authenticator {
         String version = requestContext.getMatchedAPI().getAPIConfig().getVersion();
         context = context + "/" + version;
         ResourceConfig matchingResource = requestContext.getMatchedResourcePath();
-        String httpMethod = requestContext.getMatchedResourcePath().getMethod().toString();
         SignedJWTInfo signedJWTInfo;
         try {
             signedJWTInfo = getSignedJwt(jwtToken);
         } catch (ParseException | IllegalArgumentException e) {
-            throw new SecurityException("Not a JWT token. Failed to decode the token header.", e);
+            log.error("Failed to decode the token header", e);
+            throw new APISecurityException(APIConstants.StatusCodes.UNAUTHENTICATED.getCode(),
+                    APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                    "Not a JWT token. Failed to decode the token header.", e);
         }
         JWTClaimsSet claims = signedJWTInfo.getJwtClaimsSet();
         String jwtTokenIdentifier = getJWTTokenIdentifier(signedJWTInfo);
@@ -285,7 +287,8 @@ public class JWTAuthenticator implements Authenticator {
                         CacheProvider.getGatewayJWTTokenCache().put(jwtTokenCacheKey, endUserToken);
                     } catch (JWTGeneratorException e) {
                         log.error("Error while Generating Backend JWT", e);
-                        throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                        throw new APISecurityException(APIConstants.StatusCodes.UNAUTHENTICATED.getCode(),
+                                APISecurityConstants.API_AUTH_GENERAL_ERROR,
                                 APISecurityConstants.API_AUTH_GENERAL_ERROR_MESSAGE, e);
                     }
                 }
@@ -294,7 +297,8 @@ public class JWTAuthenticator implements Authenticator {
                     endUserToken = jwtGenerator.generateToken(jwtInfoDto);
                 } catch (JWTGeneratorException e) {
                     log.error("Error while Generating Backend JWT", e);
-                    throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                    throw new APISecurityException(APIConstants.StatusCodes.UNAUTHENTICATED.getCode(),
+                            APISecurityConstants.API_AUTH_GENERAL_ERROR,
                             APISecurityConstants.API_AUTH_GENERAL_ERROR_MESSAGE, e);
                 }
             }
@@ -352,7 +356,8 @@ public class JWTAuthenticator implements Authenticator {
         } catch (EnforcerException e) {
             String message = "Error while accessing backend services for token scope validation";
             log.error(message, e);
-            throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR, message, e);
+            throw new APISecurityException(APIConstants.StatusCodes.UNAUTHENTICATED.getCode(),
+                    APISecurityConstants.API_AUTH_GENERAL_ERROR, message, e);
         }
     }
 

--- a/enforcer/src/main/resources/Dockerfile
+++ b/enforcer/src/main/resources/Dockerfile
@@ -52,7 +52,7 @@ ARG MOTD="\n\
 RUN \
     addgroup -S -g ${MG_USER_GROUP_ID} ${MG_USER_GROUP} \
     && adduser -S -u ${MG_USER_ID} -h ${MG_USER_HOME} -G ${MG_USER_GROUP} ${MG_USER} \
-    && mkdir ${MG_USER_HOME}/lib && mkdir -p ${MG_USER_HOME}/mg/security && mkdir -p ${MG_USER_HOME}/mg/logs  \
+    && mkdir ${MG_USER_HOME}/lib && mkdir -p ${MG_USER_HOME}/logs  \
     && echo '[ ! -z "${TERM}" -a -r /etc/motd ] && cat /etc/motd' >> /etc/bash.bashrc; echo "${MOTD}" > /etc/motd
 
 COPY maven/grpc_health_probe-linux-amd64 ${GRPC_HEALTH_PROBE_PATH}
@@ -67,4 +67,4 @@ COPY maven/security security
 COPY check_health.sh .
 COPY maven/LICENSE.txt .
 
-CMD java $JAVA_OPTS -Dlog4j.configurationFile="${ENFORCER_HOME}/conf/log4j2.properties" -cp "lib/*:lib/dropins/*" org.wso2.choreo.connect.enforcer.server.AuthServer
+CMD java -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath="${ENFORCER_HOME}/logs/heap-dump.hprof" $JAVA_OPTS -Dlog4j.configurationFile="${ENFORCER_HOME}/conf/log4j2.properties" -cp "lib/*:lib/dropins/*" org.wso2.choreo.connect.enforcer.server.AuthServer

--- a/enforcer/src/main/resources/Dockerfile
+++ b/enforcer/src/main/resources/Dockerfile
@@ -52,7 +52,8 @@ ARG MOTD="\n\
 RUN \
     addgroup -S -g ${MG_USER_GROUP_ID} ${MG_USER_GROUP} \
     && adduser -S -u ${MG_USER_ID} -h ${MG_USER_HOME} -G ${MG_USER_GROUP} ${MG_USER} \
-    && mkdir ${MG_USER_HOME}/lib && mkdir -p ${MG_USER_HOME}/logs  \
+    && mkdir ${MG_USER_HOME}/lib && mkdir -p ${MG_USER_HOME}/logs \
+    && chown ${MG_USER}:${MG_USER_GROUP} -R ${MG_USER_HOME} \
     && echo '[ ! -z "${TERM}" -a -r /etc/motd ] && cat /etc/motd' >> /etc/bash.bashrc; echo "${MOTD}" > /etc/motd
 
 COPY maven/grpc_health_probe-linux-amd64 ${GRPC_HEALTH_PROBE_PATH}

--- a/integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testCases/jwtValidator/JwtTestCase.java
+++ b/integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testCases/jwtValidator/JwtTestCase.java
@@ -112,4 +112,24 @@ public class JwtTestCase {
         Assert.assertTrue(response.getHeaders().containsKey("www-authenticate"),
                 "\"www-authenticate\" is not available");
     }
+
+    @Test(description = "Test to check with JWT where the JWT header cannot be parsed")
+    public void invokeJWTHeaderParseExceptionTest() throws Exception {
+
+        // Set header
+        Map<String, String> headers = new HashMap<String, String>();
+        headers.put(HttpHeaderNames.AUTHORIZATION.toString(), "Bearer " + "aaa.bbb.ccc");
+        HttpResponse response = HttpsClientRequest.doGet(Utils.getServiceURLHttps(
+                "/v2/pet/2") , headers);
+
+        Assert.assertNotNull(response);
+        Assert.assertEquals(response.getResponseCode(), HttpStatus.SC_UNAUTHORIZED,
+                "Response code mismatched");
+        Assert.assertEquals(response.getData(), "{\"error_message\":\"Unclassified Authentication Failure\"," +
+                        "\"code\":\"900900\",\"error_description\":" +
+                        "\"Not a JWT token. Failed to decode the token header.\"}",
+                "Response message mismatched");
+        Assert.assertTrue(response.getHeaders().containsKey("www-authenticate"),
+                "\"www-authenticate\" is not available");
+    }
 }

--- a/integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testCases/jwtValidator/JwtTestCase.java
+++ b/integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testCases/jwtValidator/JwtTestCase.java
@@ -125,9 +125,9 @@ public class JwtTestCase {
         Assert.assertNotNull(response);
         Assert.assertEquals(response.getResponseCode(), HttpStatus.SC_UNAUTHORIZED,
                 "Response code mismatched");
-        Assert.assertEquals(response.getData(), "{\"error_message\":\"Unclassified Authentication Failure\"," +
-                        "\"code\":\"900900\",\"error_description\":" +
-                        "\"Not a JWT token. Failed to decode the token header.\"}",
+        Assert.assertEquals(response.getData(), "{\"error_message\":\"Invalid Credentials\"," +
+                        "\"code\":\"900901\",\"error_description\":\"Not a JWT token. Failed to decode " +
+                        "the token header. Make sure you have provided the correct security credentials\"}",
                 "Response message mismatched");
         Assert.assertTrue(response.getHeaders().containsKey("www-authenticate"),
                 "\"www-authenticate\" is not available");

